Azure Policy Authoring Guide (#54)
Azure/CanadaPubSecALZ: docs/policy/authoring-guide.md

      // Start - Any custom parameters required for your policy assignment
      param ...
      // End - Any custom parameters required for your policy assignment

      // Add the GUID from the Definition ID that was gathered above
      var policyId = '<< GUID >>'

      // Add the policy set assignment name (i.e. the name of the Policy Set Name)
      var assignmentName = '<< POLICY ASSIGNMENT NAME >>'

      var scope = tenantResourceId('Microsoft.Management/managementGroups', policyAssignmentManagementGroupId)
      var policyScopedId = resourceId('Microsoft.Authorization/policySetDefinitions', policyId)

      resource policySetAssignment 'Microsoft.Authorization/policyAssignments@2020-03-01' = {

        // Set the name of the policy assignment
        // Example: name: 'pbmm-${uniqueString('pbmm-',policyAssignmentManagementGroupId)}'

        name: '<< NAME >>'

        properties: {
          displayName: assignmentName
          policyDefinitionId: policyScopedId
          scope: scope
          notScopes: [
          ]
          parameters: {
            // Add any parameters identified earlier into this section
          }

>           enforcementMode: 'Default'

Reviewer: Do we want to add comment describing allowable values for enforcement mode?

Author: Done

            },
            "EXTRA_POLICY_ASSIGNMENT_PARAMETER_NAME_1": {
                "value": "EXTRA_POLICY_ASSIGNMENT_PARAMETER_VALUE_1"
            },
            "EXTRA_POLICY_ASSIGNMENT_PARAMETER_NAME_2": {
                "value": "EXTRA_POLICY_ASSIGNMENT_PARAMETER_VALUE_2"
            }
        }
    }
    ```

    **Example:  PBMM Policy Set Parameters**

    ```json
    {
        "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
        "contentVersion": "1.0.0.0",
        "parameters": {
            "policyAssignmentManagementGroupId": {
                "value": "{{var-topLevelManagementGroupName}}"
            },
            "logAnalyticsWorkspaceId": {
                "value": "{{var-logging-logAnalyticsWorkspaceId}}"
            },
            "listOfMembersToExcludeFromWindowsVMAdministratorsGroup": {
                "value": "__tbd__implementation_specific__"
            },
            "listOfMembersToIncludeInWindowsVMAdministratorsGroup": {
                "value": "__tbd__implementation_specific__"
            }
        }
    }
    ```

#### **Step 3: Update Azure DevOps Pipeline**

>   * Edit `.pipelines/policy.yml`

Reviewer: ```suggestion
  * Edit `/.pipelines/policy.yml`
```

  * Navigate to the `BuiltInPolicyJob` Job definition
  * Navigate to the `Assign Policy Set` Step definition
  * Add the policy assignment file name (without extension) to the `deployTemplates` array parameter

#### **Step 4: Deploy built-in policy set assignment**

Execute `Azure DevOps Policy pipeline` to deploy.  The policy set assignment will be deployed to the `top level management group` (i.e. `pubsec`).

> > It takes around 30 minutes for the assignment to be applied to the defined scope. Once it's applied, the evaluation cycle begins for resources within that scope against the newly assigned policy or initiative and depending on the effects used by the policy or initiative, resources are marked as compliant, non-compliant, or exempt. A large policy or initiative evaluated against a large scope of resources can take time. As such, there's no pre-defined expectation of when the evaluation cycle completes. Once it completes, updated compliance results are available in the portal and SDKs.  See [Azure Docs for more information](https://docs.microsoft.com/azure/governance/policy/how-to/get-compliance-data).

Reviewer: ```suggestion
> It takes approximately 30 minutes for the assignment to be applied to the defined scope. Once it's applied, the evaluation cycle begins for resources within that scope against the newly assigned policy or initiative, and depending on the effects defined by the policy or initiative, resources are marked as compliant, non-compliant, or exempt. A large policy or initiative evaluated against a large scope of resources can take time. As such, there's no pre-defined expectation of when the evaluation cycle completes. Once evaluation completes, the updated compliance results are available in the portal and to SDKs.  See [Azure Docs for more information](https://docs.microsoft.com/azure/governance/policy/how-to/get-compliance-data).
```

#### **Step 5: Verify policy set assignment**

  * You can navigate to [Azure Policy Compliance][portalAzurePolicyCompliance] to verify in Azure Portal.

>   * When there are deployment errors:

Reviewer: ```suggestion
  * If there are deployment errors:
```

  
      * Navigate to [Management Groups][portalManagementGroups] in Azure Portal
      * Select the top level management group (i.e. `pubsec`)
      * Select Deployments
      * Review the deployment errors

---

### Remove built-in policy set assignment

**Steps**

* [Step 1: Remove built-in policy set assignment from Azure DevOps Pipeline](#step-1-remove-built-in-policy-set-assignment-from-azure-devops-pipeline)
* [Step 2: Remove built-in policy set assignment's IAM assignments](#step-2-remove-built-in-policy-set-assignments-iam-assignments)

#### **Step 1: Remove built-in policy set assignment from Azure DevOps Pipeline**

> * Edit `.pipelines/policy.yml`

Reviewer: ```suggestion
* Edit `/.pipelines/policy.yml`
```

    }
    ```

    The `mode` determines which resource types are evaluated for a policy definition. The supported modes are:

    * all: evaluate resource groups, subscriptions, and all resource types
    * indexed: only evaluate resource types that support tags and location

    See [Azure Policy Reference](https://docs.microsoft.com/azure/governance/policy/concepts/definition-structure#mode) for more information.

4. Edit `azurepolicy.parameters.json`.  

    Define parameters that are required by the policy definition.  Using parameters enable the policy to be used with different configuration.

    See [Azure Parameter Reference](https://docs.microsoft.com/azure/governance/policy/concepts/definition-structure#parameters) for more information.

    Example: 
    ```yml
    {
      "listOfResourceTypes": {
        "type": "Array",
        "metadata": {
          "displayName": "Resource Types",
          "description": null,
          "strongType": "resourceTypes"
        }
      }
    }
    ```

5. Edit `azurepolicy.rules.json`

    Describes the policy rule that will be evaluated by Azure Policy.  The rule can have any effect such as Audit, Deny, DeployIfNotExists.

>     See [Azure Policy docs for more information on creating custom policies](https://docs.microsoft.com/en-us/azure/governance/policy/concepts/definition-structure).

Reviewer: ```suggestion
    See [Azure Policy docs for more information on creating custom policies](https://docs.microsoft.com/azure/governance/policy/concepts/definition-structure).
```

              policyDefinitionReferenceId: toLower(replace('Deploy Log Analytics agent for Linux virtual machine scale sets', ' ', '-'))
              parameters: {
                // Set the values of parameters for each policy.
              }
            }

            // An example of a custom policy where the policy definition id is inferred from the customPolicyDefinitionMgScope variable.
            // The policy definition name is the directory name.  In this example the directory is 'LA-Logs-Diagnostic-Settings' and located in /policy/custom/definitions/policy/LA-Logs-Diagnostic-Settings
            {
              groupNames: [
                'CUSTOM'
              ]
              policyDefinitionId: extensionResourceId(customPolicyDefinitionMgScope, 'Microsoft.Authorization/policyDefinitions', 'LA-Logs-Diagnostic-Settings')
              policyDefinitionReferenceId: toLower(replace('Audit diagnostic setting', ' ', '-'))
              parameters: {
                listOfResourceTypes: {
                  value: '[parameters(\'listOfResourceTypesToAuditDiagnosticSettings\')]'
                }
              }
            }
          ]
        }
      }
    ```

> 3. Edit the JSON parameters file to define the input parameters for the Bicep template.  This parameters JSON file is used by Azure Resource Manager (ARM) for runtime inputs.

Reviewer: ```suggestion
3. Edit the JSON parameters file to define the input parameters for the Bicep template.  This JSON parameters file is used by Azure Resource Manager (ARM) for runtime inputs.
```

    **Sample Template**

    ```json
      {
          "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
          "contentVersion": "1.0.0.0",
          "parameters": {
              "policyDefinitionManagementGroupId": {
                  "value": "{{var-topLevelManagementGroupName}}"
              },
              "EXTRA_POLICY_SET_ASSIGNMENT_PARAMETER_NAME_1": {
                  "value": "EXTRA_POLICY_SET_ASSIGNMENT_PARAMETER_VALUE_1"
              },
              "EXTRA_POLICY_SET_ASSIGNMENT_PARAMETER_NAME_2": {
                  "value": "EXTRA_POLICY_SET_ASSIGNMENT_PARAMETER_VALUE_2"
              }
          }
      }
    ```

    **Example:  Log Analytics Policy Set Parameters**

    ```json
      {
          "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
          "contentVersion": "1.0.0.0",
          "parameters": {
              "policyDefinitionManagementGroupId": {
                  "value": "{{var-topLevelManagementGroupName}}"
              }
          }
      }
    ```

#### **Step 2: Create policy set assignment template**

> 1. Navigate to `policy/custom/assignments` and create two files.  Replace `POLICY_SET_ASSIGNMENT` with the name of your assignment such as `loganalytics`.

Reviewer: ```suggestion
1. Navigate to `/policy/custom/assignments` and create two files.  Replace `POLICY_SET_ASSIGNMENT` with the name of your assignment such as `loganalytics`.
```

          principalId: policySetAssignment.identity.principalId
          principalType: 'ServicePrincipal'
        }
      }

      resource policySetRoleAssignmentVirtualMachineContributor 'Microsoft.Authorization/roleAssignments@2020-04-01-preview' = {
        name: guid(policyAssignmentManagementGroupId, 'loganalytics', 'Virtual Machine Contributor')
        scope: managementGroup()
        properties: {
          roleDefinitionId: '/providers/Microsoft.Authorization/roleDefinitions/9980e02c-c2be-4d73-94e8-173b1dc7cf3c'
          principalId: policySetAssignment.identity.principalId
          principalType: 'ServicePrincipal'
        }
      }

      resource policySetRoleAssignmentMonitoringContributor 'Microsoft.Authorization/roleAssignments@2020-04-01-preview' = {
        name: guid(policyAssignmentManagementGroupId, 'loganalytics', 'Monitoring Contributor')
        scope: managementGroup()
        properties: {
          roleDefinitionId: '/providers/Microsoft.Authorization/roleDefinitions/749f88d5-cbae-40b8-bcfc-e573ddc772fa'
          principalId: policySetAssignment.identity.principalId
          principalType: 'ServicePrincipal'
        }
      }
    ```

> 3. Edit the JSON parameters file to define the input parameters for the Bicep template.  This parameters JSON file is used by Azure Resource Manager (ARM) for runtime inputs.

Reviewer: ```suggestion
3. Edit the JSON parameters file to define the input parameters for the Bicep template.  This JSON parameters file is used by Azure Resource Manager (ARM) for runtime inputs.
```

          },
          "EXTRA_POLICY_ASSIGNMENT_PARAMETER_NAME_1": {
              "value": "EXTRA_POLICY_ASSIGNMENT_PARAMETER_VALUE_1"
          },
          "EXTRA_POLICY_ASSIGNMENT_PARAMETER_NAME_2": {
              "value": "EXTRA_POLICY_ASSIGNMENT_PARAMETER_VALUE_2"
          }
        }
      }
    ```

    **Example:  Log Analytics Policy Set Parameters**

    ```json
      {
          "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
          "contentVersion": "1.0.0.0",
          "parameters": {
            "policyDefinitionManagementGroupId": {
                "value": "{{var-topLevelManagementGroupName}}"
            },
            "policyAssignmentManagementGroupId": {
                "value": "{{var-topLevelManagementGroupName}}"
            },
            "logAnalyticsWorkspaceId": {
                "value": "{{var-logging-logAnalyticsWorkspaceId}}"
            },
            "logAnalyticsResourceId": {
                "value": "{{var-logging-logAnalyticsWorkspaceResourceId}}"
            }
          }
      }
    ```

#### **Step 3: Configure Azure DevOps Pipeline**

>   * Edit `.pipelines/policy.yml`

Reviewer: ```suggestion
  * Edit `/.pipelines/policy.yml`
```

* Ensure that any new role assignments are incorporated in the policy assignment definition.
* Avoid changing the policy set definition name.  If this is required, remove the policy set assignment and its role assignments first.
* Avoid changing the policy set role assignment names.  If this is required, remove the policy set assignment and its role assignments first.

#### **Step 2: Verify policy set definition & assignment after update**

Execute `Azure DevOps Policy pipeline` to deploy the policy set definition & assignment update.

Navigate to [Azure Policy Definitions][portalAzurePolicyDefinition] and [Azure Policy Assignments][portalAzurePolicyAssignment] to verify that the policy set has been updated.

When there are deployment errors:

  * Navigate to [Management Groups][portalManagementGroups] in Azure Portal
  * Select the top level management group (i.e. `pubsec`)
  * Select Deployments
  * Review the deployment errors

---

### Remove custom policy definition

**Steps**

* [Step 1: Remove policy definition](#step-1-remove-policy-definition)

#### **Step 1: Remove policy definition**

> * Navigate to `policy/custom/definitions/policy` and identify the custom policy definition directory to delete.

Reviewer: ```suggestion
* Navigate to `/policy/custom/definitions/policy` and identify the custom policy definition directory to delete.
```

* Identify all custom policy set definitions that reference the policy (they are referenced using the directory name) and remove the references.
* Delete the custom policy definition directory in `policy/custom/definitions/policy` 
* Execute `Azure DevOps Policy pipeline` to deploy the updates.

> Azure DevOps Pipeline does not remove the custom policy definition from Azure.  It will only remove the policy definition reference from the custom policy sets that are managed through automation.  You must manually delete the policy definition using [Azure Policy Definitions][portalAzurePolicyDefinition] in Azure Portal.

---

### Remove custom policy set definition and assignment

**Steps**
  * [Step 1: Remove custom policy set definition](#step-1-remove-custom-policy-set-definition)
  * [Step 2: Remove custom policy set assignment](#step-2-remove-custom-policy-set-assignment)
  * [Step 3: Remove custom policy set from Azure DevOps Pipeline](#step-3-remove-custom-policy-set-from-azure-devops-pipeline)
  * [Step 4: Remove custom policy set assignment's IAM assignments](#step-4-remove-custom-policy-set-assignments-iam-assignments)

#### Step 1: Remove custom policy set definition

> * Navigate to `policy/custom/definitions/policyset` and delete the policy set definition Bicep template and JSON parameter files.

Reviewer: ```suggestion
* Navigate to `/policy/custom/definitions/policyset` and delete the policy set definition Bicep template and JSON parameter files.
```

#### Step 2: Remove custom policy set assignment

> * Navigate to `policy/custom/assignments` and delete the policy set assignment Bicep template and JSON parameter files.

Reviewer: ```suggestion
* Navigate to `/policy/custom/assignments` and delete the policy set assignment Bicep template and JSON parameter files.
```

#### Step 3: Remove custom policy set from Azure DevOps Pipeline

>   * Edit `.pipelines/policy.yml`

Reviewer: ```suggestion
  * Edit `/.pipelines/policy.yml`
```
